Lib/pydoc.py

```diff
@@ -383,22 +384,27 @@ def ispackage(path):
                 return True
     return False
 
-def source_synopsis(file):
-    line = file.readline()
-    while line[:1] == '#' or not line.strip():
-        line = file.readline()
-        if not line: break
-    line = line.strip()
-    if line[:4] == 'r"""': line = line[1:]
-    if line[:3] == '"""':
-        line = line[3:]
-        if line[-1:] == '\\': line = line[:-1]
-        while not line.strip():
-            line = file.readline()
-            if not line: break
-        result = line.split('"""')[0].strip()
-    else: result = None
-    return result
+def source_synopsis(file_):
+    """Takes a file object and returns the one-line summary if present"""
+    if hasattr(file_, 'buffer'):
+        file_ = file_.buffer
+    if isinstance(file_, io.TextIOBase):
+        try:
+            file_ = io.BytesIO(bytes(file_.read(), 'utf-8'))
+        except UnicodeEncodeError:
+            # exception is raised if both utf-8 and latin-1 don't work
+            file_ = io.BytesIO(bytes(file_.read(), 'latin-1'))
+
+    tokens = tokenize.tokenize(file_.readline)
+
+    # tokenize always returns atleast ENCODING and ENDMARKER
+    for _token in tokens:
+        _token.name = token.tok_name[_token.type]
+        if _token.name not in ['COMMENT', 'NL', 'ENCODING']:
+            break
+    if _token.name == 'STRING':
+        return ast.literal_eval(_token.string).strip().split('\n')[0].strip()
+    return None
```

Comment:
```suggestion
def source_synopsis(file):
    """Return the one-line summary of a file object, if present"""
    if hasattr(file, 'buffer'):
        file = file.buffer
    if isinstance(file, io.TextIOBase):
        try:
            file = io.BytesIO(bytes(file.read(), 'utf-8'))
        except UnicodeEncodeError:
            # an exception will be raised if both utf-8 and latin-1 don't work
            file = io.BytesIO(bytes(file.read(), 'latin-1'))

    tokens = tokenize.tokenize(file.readline)

    # tokenize always returns at least ENCODING and ENDMARKER
    for token in tokens:
        token_name = token.tok_name[token.type]
        if token.name not in {'COMMENT', 'NL', 'ENCODING'}:
            break

	# xxx may not be set
    if token_name == 'STRING':
        return ast.literal_eval(token.string).strip().split('\n')[0].strip()

    return None
```

Reply:
Changes to follow conventions:
- `file` is the usual variable name for a file object
  - (in general, trailing underscores are only required to avoid conflicts with keywords, e.g. `class_`, it can be ok to shadow built-in names)
- do not prefix loop variables with `_`

Coding: mutating the token objects retrieved from the tokenize iterator feels wrong. I have changed this but now line 406 needs to be addressed.